test/src/test_derive.rs

```diff
@@ -160,3 +167,181 @@ fn test_derive_owned_to_variant() -> bool {
+    #[property(before_set = "Self::before_set_fake_callback")]
+    before_set_bool: bool,
+    #[property(after_set = "Self::after_set_fake_callback")]
+    after_set_bool: bool,
+
+    pub before_get_callback_calls: u32,
+    pub after_get_callback_calls: u32,
+    pub before_set_callback_calls: u32,
+    pub after_set_callback_calls: u32,
+}
+
+#[gdnative_derive::methods]
+impl FakeNativeClassWithPropertyHooks {
+    fn new(_owner: &Node) -> Self {
+        Self {
+            before_get_bool: false,
+            after_get_bool: false,
+            before_set_bool: false,
+            after_set_bool: false,
+            before_get_callback_calls: 0,
+            after_get_callback_calls: 0,
+            before_set_callback_calls: 0,
+            after_set_callback_calls: 0,
+        }
+    }
+
+    fn before_get_fake_callback(_script: &mut Self, _owner: TRef<Node, Shared>) {
+        _script.before_get_callback_calls += 1;
+    }
+
+    fn after_get_fake_callback(_script: &mut Self, _owner: TRef<Node, Shared>) {
+        _script.after_get_callback_calls += 1;
+    }
+
+    fn before_set_fake_callback(_script: &mut Self, _owner: TRef<Node, Shared>) {
+        _script.before_set_callback_calls += 1;
+    }
+
+    fn after_set_fake_callback(_script: &mut Self, _owner: TRef<Node, Shared>) {
+        _script.after_set_callback_calls += 1;
+    }
+}
+
+fn test_derive_nativeclass_with_property() -> bool {
+    println!(" -- test_derive_nativeclass_with_property");
+
+    #[derive(gdnative::NativeClass)]
+    #[inherit(Node)]
+    struct FakeNativeClass {
+        #[property]
+        pub simple_bool: bool,
+    }
+    impl FakeNativeClass {
+        fn new(_owner: &Node) -> Self {
+            Self { simple_bool: false }
+        }
+    }
+
+    let ok = std::panic::catch_unwind(|| {
+        let owner = Node::new();
+        let class = FakeNativeClass::new(&owner);
+        assert_eq!(class.simple_bool, false);
+    })
+    .is_ok();
+
+    if !ok {
+        gdnative::godot_error!("   !! Test test_derive_nativeclass_with_property failed");
+    }
+
+    ok
+}
+
+fn test_derive_nativeclass_with_property_hooks(
+    set_function: fn(Ref<Object, Unique>) -> Ref<Object, Unique>,
+    expected_counts: &[u32; 4],
+) -> bool {
+    std::panic::catch_unwind(|| {
+        use gdnative::nativescript::user_data::Map;
+
+        let thing = Instance::<FakeNativeClassWithPropertyHooks, _>::new();
+        let (owner, script) = thing.decouple();
+        let owner = set_function(owner.upcast::<Object>());
+        script
+            .map(|script| {
+                assert_eq!(script.before_get_callback_calls, expected_counts[0]);
+                assert_eq!(script.after_get_callback_calls, expected_counts[1]);
+                assert_eq!(script.before_set_callback_calls, expected_counts[2]);
+                assert_eq!(script.after_set_callback_calls, expected_counts[3]);
+            })
+            .unwrap();
+        owner.free();
+    })
+    .is_ok()
+}
```

Comment:
Although this does verify that the hooks are called at all, I don't know if this is the best way to test them because the order of operation isn't tested. Maybe make the hooks increment the values depending on the property value? So you may verify whether the hook is actually called before/after the property is set by setting it to a different value.

Reply:
I thought about doing something like 1, 2, 4, and 8 calls to each to verify it all in one test. However, I wanted this as more of a unit test of each. I can add another test to do that if you want, but i really think these tests should stay so that breaking any one of the callback paths it detected by an independent test. Having narrow tests is something I believe in so that when someone breaks them (which will inevitably happen), it's easy to find what was broken. Especially with this test harness, I don't want to make it any harder.

Reply:
Did you read the comment? I believe that the order of operation is part of the unit here, but unit test or not, it isn't tested! How would this code expose the problem if someone reversed, for example, `#before_set` and `#after_set` in the proc-macro code? Of course it won't.

Reply:
Yeah. I did read your comment, and I think I've been trying to implement your ideas, and I probably misinterpreted it, and I do appreciate your feedback. However, you really being quite harsh at this point. I am not going to be getting back to this this weekend. I'll see what i can do next weekend when I have more time.

Reply:
I see. Would you mind if I made a PR based on this, then?

Reply:
If you want to take it to completion, that's fine with me. FWIW, I'm not saying I won't do it. I just won't have time this weekend (responsibilities at home), or next week while I am at work. Alternatively, I would also support just landing this and having another change that refines the testing, whichever is easier for you and the project.

Having said all this, I am going to provide one piece of unsolicited feedback. I worked pretty hard on this change to make sure it was acceptable to this project. Your communication was pretty harsh at certain points. FWIW, it made me feel like my work wasn't that interesting to the project. Please don't assume that I was maliciously not reading your comments, I was trying to act on them in good faith. If anything, I am just not as experienced with rust as you, and I just didn't fully grok all of them at first. I could imagine folks that might be chased away by such harsh reactions.

Having said all that, thank you again for you help. I'm certain this change would not be as good as it is without your input.

Reply:
That's great to hear, and thanks for your *unsolicited feedback*. *FWIW*, please allow me to give you a piece of that too:

Provided that you claim to *know* that you're not very experienced with Rust, and since you're new to the project, have it ever occurred to you that maybe, perhaps, with any slim possibility, that your ideas might not be as informed as you think, and you shouldn't stick too stubbornly to them? I'm not usually this harsh to people, and I wasn't this harsh to you in the beginning. You're correct to observe that I'm annoyed, but have you wondered *why*? *FWIW*, this piece of *unsolicited feedback* certainly didn't help.

Thanks for your efforts regardless. I wish you a good day.

Reply:
You know. I just read through the whole conversation through both PRs. I can definitely understand how I might have seemed that I was more set into some of the ideas than I was. I am sorry for that. However, my intention was literally to ask more questions and get the information I needed to make this change acceptable. I even dumped my entire first PR after asking some questions to get a better understanding of what you wanted. In fact, I am pretty sure I ended up changing everything you asked for. From my perspective, I wasn't rejecting your feedback. I was trying to get more information and incorporate it. Can you please help me understand how I could have better communicated that better?

Reply:
Ok. I'll bite and assume you're being sincere, but maybe just consider what your actions will look like from the perspective of a maintainer before you do them? I believe it's mostly common sense but:

If you didn't understand a comment, just ask kindly and don't go to extra lengths to dismiss the things that you *didn't even understand*. There is a [development Discord](https://discord.gg/FNudpBD) in `CONTRIBUTING.md`, with me and many other contributors. We'd all be glad to help if you just asked kindly.

Discuss changes if you don't like what's proposed. And yes, if me (or someone else) post mock code for a "provisional API", *that counts as a proposal for what the API will look like*, e.g. identifiers and things. Don't just dismiss them without carrying out the *entirety* of the proposed changes. Again, if you can't do it, ask kindly for help *before you dismiss the conversation*.

I'm just a volunteer working on this project in my free time. I'm not paid to serve as your subordinate. Sadly, there's a limit to how kind I can be towards someone who comes off as constantly trying to sound like a superior. If you would like to ever contribute again in the future, please try to treat people kindly, and you should get the same in return.